src/Entity/Int32EntityId.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+namespace Wikibase\DataModel\Entity;
+
+use InvalidArgumentException;
+
+/**
+ * Interface for EntityIds that can be converted to a unique positive 32 bit integer in the range
+ * [1..2147483647], and back from the entity type and the number. Entity types that do not meet this
+ * criteria should not implement this interface.
+ *
+ * Entity types are not required and not guaranteed to implement this interface. Use the full string
+ * serialization whenever you can and avoid using numeric IDs.
+ *
+ * @since 6.1
+ *
+ * @license GPL-2.0+
+ * @author Thiemo Mättig
+ */
+interface Int32EntityId {
+
+	/**
+	 * @since 6.1
+	 *
+	 * @param string $idSerialization
+	 *
+	 * @throws InvalidArgumentException When the provided serialization does not allow getNumericId
+	 *  to return a unique positive 32 bit integer.
+	 */
+	public function __construct( $idSerialization );
+
+	/**
+	 * @since 6.1
+	 *
+	 * @return int Guaranteed to be a unique positive 32 bit integer.
+	 */
+	public function getNumericId();
+
+	/**
+	 * @since 6.1
+	 *
+	 * @param int|float|string $numericId
```

Comment:
What should newFromNumber( '020' ) return? "X020" or "X20" or "X16"?
That's 'X' . $numericId  vs.  'X' . (int)$numericId  vs.  'X' . intval($numericId, 0).  
I checked again now, and it seems the case of X16 isn't as relevant as I thought, since intval doesn't do "automatic octal" per default. But the spec should protect against "X020".